### PR TITLE
soc: riscv: sifive-freedom: Specify clocks in CPU DTS and allow custom FE310 CPU clock rate.

### PR DESCRIPTION
**Pull request by snematbakhsh**
tlclk, which is the input clock for SPI, I2C, PWM, etc. is (mostly) a fixed
ratio of coreclk. Rather than specify input clock for each peripheral
individually, instead specify CPU clock (coreclk) and derive tlclk from
it.

This will enable easier support for non-default coreclk on fe310 in a
follow-up CL.

Signed-off-by: Shawn Nematbakhsh <shawn@rivosinc.com>

### CODE REVIEW

File: dts/bindings/riscv/riscv,cpus.yaml [hunk 1]
```diff
@@ -24,8 +24,3 @@ properties:
       - rv32imafcb
       - rv64imac
       - rv64imafdc
-
-  riscv,pmpregions:
-    description: Physical Memory Protection (PMP)
-    required: true
-    type: int
```
Comment (dcpleung) [inline dts/bindings/riscv/riscv,cpus.yaml:31]: Why are you removing this? This removal does not seem to be related to this PR.

Comment (snematbakhsh) [reply dts/bindings/riscv/riscv,cpus.yaml]: The property was marked as required but not defined or used anywhere. Is changing to "required: false" preferred instead of removing?

Comment (dcpleung) [reply dts/bindings/riscv/riscv,cpus.yaml]: A separate commit explaining the removal would be better.

Comment (snematbakhsh) [reply dts/bindings/riscv/riscv,cpus.yaml:31]: Done.

File: drivers/spi/spi_sifive.c [hunk 2]
```diff
@@ -259,7 +259,7 @@ static struct spi_driver_api spi_sifive_api = {
 	}; \
 	static struct spi_sifive_cfg spi_sifive_cfg_##n = { \
 		.base = DT_INST_REG_ADDR_BY_NAME(n, control), \
-		.f_sys = DT_INST_PROP(n, clock_frequency), \
+		.f_sys = SIFIVE_TLCLK_HZ, \
 	}; \
```
Comment (katsuster) [inline drivers/spi/spi_sifive.c:263]: Hi. In FU740, SPI and other peripherals are not driven by tlclk. This name is not suitable...

Specification of FU740 "7.1 Clocking" said that
`The hfpclkpll generates the clock for peripherals such as SPI, UART, GPIO, I2C, and PWM.`
This hfpclkpll is different PLL from tlclk.

Comment (snematbakhsh) [reply drivers/spi/spi_sifive.c:263]: Thanks for checking it. Here is a citation from the the same FU740 specification, "19.4 Serial Clock Divisor Register (sckdiv)".

`The relationship between the input clock and SCK is given by the following formula:`
`fsck = (fin) / 2 * (div + 1)`
`The input clock is the bus clock tlclk.`

And, there is a similar note regarding tlclk being the input clock for the UART peripheral (16.9). But I do see that PCLK is specified as the input in Linux (arch/riscv/boot/dts/sifive/fu740-c000.dtsi).

So I suspect you are correct, but it's difficult to resolve these conflicting statements in the specification. I will do some testing on an FU740 board to confirm.

**Review by mbolivar-nordic — COMMENTED** [review]
No objections from me if the RISC-V maintainers are happy with this approach

**Comment by fkokosinski** [comment]
Please rebase.

File: drivers/spi/spi_sifive.c (hunk 2 above)
Comment (snematbakhsh) [reply drivers/spi/spi_sifive.c:263]: I tested on an FU740 board and confirmed that UART is functional with the existing 125125000Hz clock setting. "_The input clock is the bus clock tlclk._" - this is an erroneous statement for FU740. I will revise the patch accordingly.

**Comment by snematbakhsh** [comment]
Updated patch set:
- Rebased to latest main.
- Clocks are specified differently for FU740 vs other chips (thanks to [katsuster](https://github.com/katsuster) for pointing out the problem).
- Added third patch which allows changing FE310 CPU clock from defaut.

**Review by fkokosinski — DISMISSED** [review]
LGTM

File: dts/riscv/riscv32-fe310.dtsi [hunk 3]
```diff
@@ -1,29 +1,36 @@
...
 	cpus {
 		#address-cells = <1>;
 		#size-cells = <0>;
 		cpu: cpu@0 {
-			clock-frequency = <0>;
-			compatible = "sifive,rocket0", "riscv";
 			device_type = "cpu";
-			i-cache-block-size = <64>;
-			i-cache-sets = <128>;
-			i-cache-size = <16384>;
```
Comment (tgorochowik) [inline dts/riscv/riscv32-fe310.dtsi:19]: Removal of these properties  (and some below) also seems unrelated to the commit message. Can we split it out and add a proper explanation for the removal?

**Review by fkokosinski — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
@kgugala could you take a look?

**Review by katsuster — APPROVED** [review]

File: soc/riscv/riscv-privilege/sifive-freedom/soc.h [hunk 4]
```diff
@@ -23,7 +23,7 @@
 #define SIFIVE_PINMUX_PINS            32
 
 /* Clock controller. */
-#define PRCI_BASE_ADDR               0x10008000
+#define PRCI_BASE_ADDR           0x10008000
```
Comment (kgugala) [inline soc/riscv/riscv-privilege/sifive-freedom/soc.h:26]: looks like this is accidental change, please remove it

**Review by kgugala — APPROVED** [review]
One small formatting comment, otherwise LGTM

File: soc/riscv/riscv-privilege/sifive-freedom/soc.h (hunk 4 above)
Comment (carlescufi) [reply soc/riscv/riscv-privilege/sifive-freedom/soc.h:26]: @snematbakhsh please send a separate PR to fix this.

Comment (snematbakhsh) [reply soc/riscv/riscv-privilege/sifive-freedom/soc.h:26]: Sorry, it's my mistake. I removed those spaces to align with the definitions below, but in fact the alignment was already fine with the definitions above (and definitions below use tabs, so I definitely made things worse). PR #44560 sent.
